File: 프로그래머스/Python/2단계/[2022 KAKAO BLIND RECRUITMENT] - 주차 요금 계산.py
"""
주차장의 요금표와 차량이 들어오고(입차) 나간(출차) 기록이 주어졌을 때, 차량별로 주차 요금을 계산하려고 합니다. 

- 어떤 차량이 입차된 후에 출차된 내역이 없다면, 23:59에 출차된 것으로 간주합니다.
- 00:00부터 23:59까지의 입/출차 내역을 바탕으로 차량별 누적 주차 시간을 계산하여 요금을 일괄로 정산합니다.
- 누적 주차 시간이 기본 시간이하라면, 기본 요금을 청구합니다.
- 누적 주차 시간이 기본 시간을 초과하면, 기본 요금에 더해서, 초과한 시간에 대해서 단위 시간 마다 단위 요금을 청구합니다.
    -초과한 시간이 단위 시간으로 나누어 떨어지지 않으면, 올림합니다.
    -⌈a⌉ : a보다 작지 않은 최소의 정수를 의미합니다. 즉, 올림을 의미합니다.

주차 요금을 나타내는 정수 배열 fees, 자동차의 입/출차 내역을 나타내는 문자열 배열 records가 매개변수로 주어집니다. 
차량 번호가 작은 자동차부터 청구할 주차 요금을 차례대로 정수 배열에 담아서 return 하도록 solution 함수를 완성해주세요.
"""

# 내 풀이
from datetime import datetime
import math

def solution(fees, records):
    result = []  # 정답을 담는 배열
    state = {}   # 차량의 입차 시간을 담는 딕셔너리
    car_dict = {} # 차량의 누적 주차시간을 담는 딕셔너리
    
    for i in records:
        time, num, re = i.split()  # 시간, 차량 번호, 상태 저장
        if num not in car_dict:  # 만약에 누적 주차시간에 차량이 등록되지 않았다면 
            car_dict[num] = 0  # 차량 추가 후 초기화 
        if re == 'IN': # 입차
            state[num] = time  # 입차 시간 등록
        elif re == 'OUT':  # 출차
            in_time = datetime.strptime(state[num], "%H:%M")  # 들어온 시간
            out_time = datetime.strptime(time, "%H:%M")   # 나간 시간
            p_time = ((out_time - in_time).seconds)//60  # 시간차를 초로 변환한 후, 분으로 바꿈
            car_dict[num] += int(p_time) # 누적시간 저장
            state.pop(num, None)  # 차량 출차 처리

    for i in state:  # 출차하지 않은 차들
        in_time = datetime.strptime(state[i], "%H:%M")
        last_time = datetime.strptime('23:59', "%H:%M")  # 마지막 시간을 23:59로 처리
        p_time = ((last_time - in_time).seconds)//60
        car_dict[i] += int(p_time)
        
    for k in sorted(car_dict):  # 차량 오름차순 정렬
        fee = fees[1]  # 기본 요금 저장
        if car_dict[k] <= fees[0]:  # 기본 시간 이하라면
            result.append(fee)  # 그대로 저장
        else:  # 기본시간 이상이라면
            fee += math.ceil((car_dict[k] - fees[0])/fees[2]) * fees[3]  # 요금 계산
            result.append(fee) # 저장
            
    return result



# 다른사람 풀이
from collections import defaultdict
from math import ceil

class Parking:
    def __init__(self, fees):
        self.fees = fees
        self.in_flag = False
        self.in_time = 0
        self.total = 0

    def update(self, t, inout):
        self.in_flag = True if inout=='IN' else False
        if self.in_flag:  self.in_time = str2int(t)
        else:             self.total  += (str2int(t)-self.in_time)

    def calc_fee(self):
        if self.in_flag: self.update('23:59', 'out')
        add_t = self.total - self.fees[0]
        return self.fees[1] + ceil(add_t/self.fees[2]) * self.fees[3] if add_t >= 0 else self.fees[1]

def str2int(string):
    return int(string[:2])*60 + int(string[3:])

def solution(fees, records):
    recordsDict = defaultdict(lambda:Parking(fees))
    for rcd in records:
        t, car, inout = rcd.split()
        recordsDict[car].update(t, inout)
    return [v.calc_fee() for k, v in sorted(recordsDict.items())]

File: 프로그래머스/JAVA/2단계/[2022 KAKAO BLIND RECRUITMENT] - 주차 요금 계산.java
/*
https://school.programmers.co.kr/learn/courses/30/lessons/92341
: 차량별 총 주차 시간을 구한 뒤 요금 정책에 따라 계산하고, 차량 번호를 오름차순으로 출력하는 문제

[접근]
"상태(Map) + 누적(Map) 분리"

1. 차량 상태 관리(입차/출차)
  - inMap: 현재 주차 중인 차량의 입차 시간
  - totalMap: 차량별 누적 주차 시간
  
  👉 상태와 결과를 분리하는 것이 핵심이다.

2. 기록을 순회하면서 시간 누적
  - IN ⇢ 입차 시간 저장
  - OUT ⇢ (현재 시간 - 입차 시간) 누적

3. 출차 안 된 차량 처리
  - 문제에서 가장 중요한 조건
  - 끝까지 안 나가면 23:59 기준으로 계산

4. 차량 번호 기준 정렬
  - Map은 정렬 안 되므로 따로 정렬 필요

5. 요금 계산 
  - 기본 시간 이하면 기본 요금
  - 초과 시 올림 계산
*/

import java.util.*;

class Solution {
    public int[] solution(int[] fees, String[] records) {
        Map<String, Integer> inMap = new HashMap<>();   // 입차 시간
        Map<String, Integer> totalMap = new HashMap<>(); // 누적 시간
        
        int base_time = fees[0];
        int base_fee  = fees[1];
        int per_time  = fees[2];
        int per_fee   = fees[3];
        
        for (String record: records){
            String[] info = record.split(" ");
            
            String time = info[0];
            String car = info[1];
            String type = info[2];
            
            int minute = convertToMin(time);
            
            if (type.equals("IN")){
                inMap.put(car, minute);
            }else {
                int inTime = inMap.remove(car); // Map.remove(): 값을 꺼내면서 삭제
                totalMap.put(car, totalMap.getOrDefault(car, 0) + (minute - inTime)); // getOrDefault: key 없으면 기본값 반환 (Null 체크 필요 없음)
            }
        }
        for (String car: inMap.keySet()){ // Map.keySet(): key만 가져올 때 사용
            int inTime = inMap.get(car);
            int minute = 23*60+59;
            totalMap.put(car, totalMap.getOrDefault(car, 0) + (minute - inTime));
        }
                        
        List<String> cars = new ArrayList<>(totalMap.keySet());
        Collections.sort(cars);
        
        int[] answer = new int[cars.size()];
        int idx = 0;
        
        for (String car: cars){
            int time = totalMap.get(car);
            int fee = base_fee;
            
            if (time > base_time){
                fee += Math.ceil((time - base_time) / (double) per_time) * per_fee; // Math.ceil + 형변환: (double) 없으면 정수 나눗셈 됨
            }
            answer[idx++] = fee;
        }
        return answer;
    }
    
    public int convertToMin(String time){
        String[] tmp = time.split(":");
        
        return Integer.parseInt(tmp[0]) * 60 + Integer.parseInt(tmp[1]);
    }
}
